### PR TITLE
Fix 711 in FIX

### PR DESCRIPTION
**Pull request by foroughgoudarzi**

### CODE REVIEW

File: src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java [hunk 1]
```diff
@@ -144,6 +147,24 @@ public long pauseMillisBeforeReconnect() {
...
     }
 
+    public long minTimeSec() {
+        return minTimeSec;
+    }
+
+    public AlwaysStartOnPrimaryConnectionStrategy minTimeSec(long minTimeSec) {
+        this.minTimeSec = minTimeSec;
+        return this;
+        }
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java:157]: formatting

File: src/main/java/net/openhft/chronicle/network/ConnectionStrategy.java [hunk 2]
```diff
@@ -183,5 +183,12 @@ default void throwExceptionIfClosed() throws IllegalStateException {
             throw new ClosedIllegalStateException("Closed");
         }
     }
-    
+
+     long minTimeSec();
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/network/ConnectionStrategy.java:187]: I don't believe it is necessary to add these to `ConnectionStrategy`

File: src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java [hunk 3]
```diff
@@ -51,6 +51,9 @@
     private int socketConnectionTimeoutMs = Integer.getInteger("connectionStrategy.socketConnectionTimeoutMs", 1);
     private long pauseMillisBeforeReconnect = Integer.getInteger("connectionStrategy.pauseMillisBeforeReconnect", 500);
     private ClientConnectionMonitor clientConnectionMonitor = new VanillaClientConnectionMonitor();
+    private long minTimeSec = Integer.getInteger("connectionStrategy.pause.min.secs", 5);
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java]: I think you only need to change this file?

Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java]: ah or else maybe just default get methods in `ConnectionStrategy` like we do for `pauseMillisBeforeReconnect`

Comment (JanStureNielsen) [inline src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java]: This property name introduces yet another naming scheme to the configuration properties; consider using one of the existing, e.g., `connectionStrategy.minimumPauseSeconds` and `connectionStrategy.maximumPauseSeconds`?

And perhaps open a new issue to address the `Ms` vs `Millis` and lack of units in the other property names.

File: src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java [hunk 4]
```diff
@@ -144,6 +147,24 @@ public long pauseMillisBeforeReconnect() {
         return pauseMillisBeforeReconnect;
     }
 
+    public long minTimeSec() {
+        return minTimeSec;
+    }
+
+    public AlwaysStartOnPrimaryConnectionStrategy minTimeSec(long minTimeSec) {
```
Comment (JanStureNielsen) [inline src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java]: Perhaps more descriptively: `minimumPauseSeconds`?

File: src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java [hunk 5]
```diff
@@ -144,6 +147,24 @@ public long pauseMillisBeforeReconnect() {
         return pauseMillisBeforeReconnect;
     }
 
+    public long minTimeSec() {
```
Comment (JanStureNielsen) [inline src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java]: `minimumPauseSeconds`?

File: src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java [hunk 6]
```diff
@@ -144,6 +147,24 @@ public long pauseMillisBeforeReconnect() {
...
+        return minTimeSec;
+    }
+
+    public AlwaysStartOnPrimaryConnectionStrategy minTimeSec(long minTimeSec) {
+        this.minTimeSec = minTimeSec;
+        return this;
+        }
+
+    public AlwaysStartOnPrimaryConnectionStrategy maxTimeSec(long maxTimeSec) {
```
Comment (JanStureNielsen) [inline src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java]: `maximumPauseSeconds`?

File: src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java (hunk 3 above)
Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java]: @JanStureNielsen these property names are legacy, and in use by customers

File: src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java [hunk 7]
```diff
@@ -144,6 +147,24 @@ public long pauseMillisBeforeReconnect() {
...
+        this.maxTimeSec = maxTimeSec;
+        return this;
+        }
+
+    public long maxTimeSec() {
```
Comment (JanStureNielsen) [inline src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java]: `maximumPauseSeconds`?

File: src/main/java/net/openhft/chronicle/network/ConnectionStrategy.java [hunk 8]
```diff
@@ -183,5 +183,12 @@ default void throwExceptionIfClosed() throws IllegalStateException {
             throw new ClosedIllegalStateException("Closed");
         }
     }
-    
+
+    default long minTimeSec() {
+        return 5;
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/network/ConnectionStrategy.java]: so I think these are the only methods that need to be added, and they should refer to the properties e.g. `Integer.getInteger("connectionStrategy.pause.min.secs", 5)`

I don't believe we need any changes to AlwaysStart... and FatalFailre...

File: src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java (hunk 3 above)
Comment (JanStureNielsen) [reply src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java]: Oh, bummer -- OT: is there support for alias so the code-base can be consistently named while retaining the legacy configuration name and presenting a more consistent naming scheme?

**Comment by foroughgoudarzi** [comment]
updated the variable names and converted the changes of AlwaysStart... and FatalFailre... and moved the system properties to the ConnectionStrategy interface. 
I inferred from the ticket title that we want to configure these variables from the YAML so I defined the system properties in AlwaysStart... and FatalFailre...  first.

File: src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java [hunk 9]
```diff
@@ -45,11 +45,11 @@
 public class AlwaysStartOnPrimaryConnectionStrategy extends AbstractMarshallableCfg implements ConnectionStrategy {
 
     private static final Logger LOG = LoggerFactory.getLogger(AlwaysStartOnPrimaryConnectionStrategy.class);
-
-    private int tcpBufferSize = Jvm.getInteger("tcp.client.buffer.size", TCP_BUFFER);
-    private int pausePeriodMs = Jvm.getInteger("client.timeout", 500);
-    private int socketConnectionTimeoutMs = Jvm.getInteger("connectionStrategy.socketConnectionTimeoutMs", 1);
-    private long pauseMillisBeforeReconnect = Jvm.getInteger("connectionStrategy.pauseMillisBeforeReconnect", 500);
+    private final transient AtomicBoolean isClosed = new AtomicBoolean(false);
+    private int tcpBufferSize = Integer.getInteger("tcp.client.buffer.size", TCP_BUFFER);
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java]: why change from Jvm.getInteger to Integer.getInteger?

**Comment by JerryShea** [comment]
> updated the variable names and converted the changes of AlwaysStart... and FatalFailre... and moved the system properties to the ConnectionStrategy interface. I inferred from the ticket title that we want to configure these variables from the YAML so I defined the system properties in AlwaysStart... and FatalFailre... first.

Ah yes.... sorry... You have a good point there. We will need the fields defined in `AlwaysStart...` so we can set with YAML

File: src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java (hunk 3 above)
Comment (RobAustin) [reply src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java]: > This property name introduces yet another naming scheme to the configuration properties; consider using one of the existing, e.g., `connectionStrategy.minimumPauseSeconds` and `connectionStrategy.maximumPauseSeconds`?
> 
> And perhaps open a new issue to address the `Ms` vs `Millis` and lack of units in the other property names.

I agree - these new suggested method names are better

**Comment by minborg** [comment]
Generally, variables involving time are named `fooS`, `barMs` and `bazzNs` using the SI time units with capital letters. Plural forms should be avoided.

File: src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java (hunk 3 above)
Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java]: > > This property name introduces yet another naming scheme to the configuration properties; consider using one of the existing, e.g., `connectionStrategy.minimumPauseSeconds` and `connectionStrategy.maximumPauseSeconds`?
> > And perhaps open a new issue to address the `Ms` vs `Millis` and lack of units in the other property names.
> 
> I agree - these new suggested method names are better

Yes but that is changing property names that customers are using. If we change then we need to support the old for a period. If we are going to change we should review all the system properties and consistentise them

File: src/main/java/net/openhft/chronicle/network/ConnectionStrategy.java (hunk 8 above)
Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/network/ConnectionStrategy.java]: These should refer to the system properties. AlwaysStartOnPrimaryConnectionStrategy can refer to these thus ensuring system properties are in one place

File: src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java (hunk 3 above)
Comment (alamar) [reply src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java]: Please always use Jvm.getInteger() because it allows code user to set those properties in advance in controlled fashion.

Comment (foroughgoudarzi) [reply src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java]: Thanks, already did, see the last changes.

File: src/main/java/net/openhft/chronicle/network/ConnectionStrategy.java (hunk 8 above)
Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/network/ConnectionStrategy.java]: @foroughgoudarzi please see my last comment

File: src/main/java/net/openhft/chronicle/network/ConnectionStrategy.java [hunk 10]
```diff
@@ -40,6 +40,7 @@
 
 public interface ConnectionStrategy extends Marshallable, java.io.Closeable {
 
+
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/network/ConnectionStrategy.java]: added unnecessary change

File: src/main/java/net/openhft/chronicle/network/connection/FatalFailureConnectionStrategy.java [hunk 11]
```diff
@@ -202,4 +202,5 @@ public void readMarshallable(@NotNull WireIn wire) throws IORuntimeException {
         super.readMarshallable(wire);
         init();
     }
+
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/network/connection/FatalFailureConnectionStrategy.java]: don't need this extra line

File: src/main/java/net/openhft/chronicle/network/ConnectionStrategy.java (hunk 8 above)
Comment (foroughgoudarzi) [reply src/main/java/net/openhft/chronicle/network/ConnectionStrategy.java]: All system properties are in one place (AlwaysStartOnPrimaryConnectionStrategy). Please let me know where would you like to define these two system properties? ConnectionStrategy is an interface and its variable would be final and not configurable. I have added these system properties the same way that pauseMillisBeforeReconnect has been defined.

Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/network/ConnectionStrategy.java]: We have to have the system property in the interface so that all implementations (even customer impls) use it, and we don't get behaviour changes
```
default long minTimeSec() {
        return Integer.getInteger("connectionStrategy.pause.min.secs", 5);
}
```
and in AlwaysStart... you can refer to the interface to default the field so we minimise duplication
```
private long minPauseSec = ConnectionStrategy.super.minTimeSec();
```
Don't forget to squash merge all these commits!

File: src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java (hunk 7 above)
Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/network/AlwaysStartOnPrimaryConnectionStrategy.java]: Yeah agree that we should not call them maxTimeSec - max time of what? Should be maxPauseSec IMO

**Review by JerryShea — APPROVED** [review]
